man/lsxid.8
.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.41.2.
.TH LSXID "8" "May 2013" "lsxid  -- shows the context which is associated to a file" "System Administration"
.SH NAME
lsxid \- shows the context which is associated to a file
.SH SYNOPSIS
.B lsxid
[\fI-Radnx\fR] [\fI--\fR] \fI<file>*\fR
.SH DESCRIPTION
.IP
Options:
.TP
\fB\-R\fR
\&...  recurse through directories
.TP
\fB\-a\fR
\&...  display files starting with '.' also
.TP
\fB\-d\fR
\&...  list directories like other files instead of
.IP
listing their content
.TP
\fB\-n\fR
\&...  do not try to do xid \-> vserver\-name mapping
.TP
\fB\-x\fR
\&...  do not cross filesystems
.PP
Please report bugs to vserver@list.linux\-vserver.org
.PP
This program is part of util\-vserver
.SH COPYRIGHT
Copyright \(co 2004 Enrico Scholz
This program is free software; you may redistribute it under the terms of
the GNU General Public License.  This program has absolutely no warranty.


NEWS
version 0.30.215
================

	- added support for using libnss instead of beecrypt for
	vhashify's hash calculation.  libnss has bad SHA1 performance on
	i386, but is a) maintained and b) gives better performance with
	all other hashes and on x86_64.

	- added tunctl, a small program to setup tun/tap interfaces, with
	support in the configuration.

	- vattribute and nattribute now support a --get operation, which
	parses the flags and capabilities into a human-readable form.

	- vspace supports all the spaces currently available.

	- vmemctrl controls the OOM bias of context.

	- dynamic contexts are unbroken.

	- hopefully all rpm-fake races have been fixed.

	- flock has been replaced by lockf, which works on NFS.

	- the clone build method now supports exclude lists.

	- vmount handles mounting into a guest.

	- various fixes for the Gentoo-scripts, thanks to Benedikt Böhm.

	- support for yum's persistdir has been added.

	- per-guest device maps are now supported.

	- vserver ... build now sets the barrier on the parent directory.

	- fsck is now supported for the guest's fstab.

	- symlinks are now handled on all levels in namespace cleanup.

	- the debootstrap build method now supports debootstrap 1.0.8.

	- a documentation update, thanks to Micah Anderson.

	- use numeric uids/gids for the template build method.

	- add support for the alpine and f8 distributions.
	
version 0.30.214
================
	- patches for yum-3.2.

	- various fixes to allow for *SuSE guests.

	- added Fedora 7.

	- replace file with h2ext for build -m template.

	- add exclude list support for vclone.

	- include debootstrap scripts for a lot of distributions.

	- fixes to allow the Gentoo initscripts to be used with baselayout 2.

	- remove disk limit caches on vserver ... delete.

	- added support for the new APIs in 2.2 and 2.3, in particular the
	netv2 API, allowing native IPv6 support.

	- added support for a separate filesystem tag.

version 0.30.213
================
	- added support for the device mapping feature present in development
	kernels.

	- further improved Gentoo support, again thanks to Benedikt Böhm.

	- support persistent contexts.

	- add another initscript that creates directories, configures vshelper
	as well as kills all contexts.

	- new vserver-stat implementation, which uses the accounting APIs
	available in Linux-VServer 2.2.

	- added support for setting sysctl values in the guest configuration.

	- added a clone build method, which simply links already unified files.

	- implement dynamic contexts in userspace, generating a static one
	during vserver ... build.

	- fix the endless loop in vlogin.

	- fix Ubuntu builds.

	- improved support for Debian Etch, thanks to Benedikt Böhm.

	- added CentOS 5.

	- use the state change helper to try to cleanup after us if the guest
	exited on its own.

version 0.30.212
================
	- support for the Linux-VServer 2.1 API has been implemented.

	- the scheduler configuration has been split up into multiple files.

	- namespace cleanup is enabled by default. This can be disabled by
	touch ${sysconfdir}/vservers/.defaults/nonamespace-cleanup

	- all unset ulimits are now reset to infinity, so limits are no longer
	inherited from the host.

	- much improved Gentoo support, thanks to Benedikt Böhm.

version 0.30.211
================

	- novlandev is now the default. To reenable the old behaviour of
	creating VLAN interfaces where used, touch
	${sysconfdir}/vservers/.defaults/interfaces/vlandev.

	- vshelper now works correctly with the kernel legacy support turned
	off.

	- vlogin provides a terminal proxy which allocates a new psuedo-tty
	on the inside when you enter a guest.

	- and lots of other features/fixes, see ChangeLog.


version 0.30.210
================

	- LEGACY BREAKAGE: by default, only recent kernel APIs will be
	supported so you have to make sure to use at least the vs2.0.1
	kernel patch. With older kernels, you have to call 'configure'
	with '--enable-apis=NOLEGACY'. Some programs will not work anymore
	with the default settings; these ones are:

	     + capchroot
	     * chcontext-context
	     * reducecap

	- dlimit support was integrated into the standard vserver
	operations. Thanks to Daniel Hokka Zakrisson for his patch.

	- recent kernel API is now used for setting up the network
	contexts. So it should be now possible to use a kernel compiled
	without CONFIG_VSERVER_LEGACYNET. Thanks to Daniel Hokka Zakrisson
	for his patch.  .


version 0.30.209
================

	- large revamping...
	
	
version 0.23.6
==============

	- fixed '--level' option on 'vserver ... chkconfig'


src/lsxid.c
// $Id$    --*- c -*--

// Copyright (C) 2004 Enrico Scholz <enrico.scholz@informatik.tu-chemnitz.de>
//  
// This program is free software; you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation; version 2 of the License.
//  
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//  
// You should have received a copy of the GNU General Public License
// along with this program; if not, write to the Free Software
// Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.


#ifdef HAVE_CONFIG_H
#  include <config.h>
#endif

#include "fstool.h"
#include "util.h"

#include <ensc_vector/vector.h>
#include <lib/vserver.h>
#include <lib/vserver-internal.h>

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <sys/param.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <assert.h>

#define ENSC_WRAPPERS_STDLIB	1
#include <wrappers.h>

struct option const
CMDLINE_OPTIONS[] = {
  { "help",     no_argument,  0, CMD_HELP },
  { "version",  no_argument,  0, CMD_VERSION },
  { 0,0,0,0 }
};

char const		CMDLINE_OPTIONS_SHORT[] = "Radnx";

struct XidNameMapping {
    xid_t		xid;
    char const *	name;
};

static struct Vector	xid_name_mapping;

void
showHelp(int fd, char const *cmd, int res)
{
  WRITE_MSG(fd, "Usage:  ");
  WRITE_STR(fd, cmd);
  WRITE_MSG(fd,
	    " [-Radnx] [--] <file>*\n\n"
	    " Options:\n"
	    "   -R  ...  recurse through directories\n"
	    "   -a  ...  display files starting with '.' also\n"
	    "   -d  ...  list directories like other files instead of\n"
	    "            listing their content\n"
	    "   -n  ...  do not try to do xid -> vserver-name mapping\n"
	    "   -x  ...  do not cross filesystems\n\n"
	    "Please report bugs to " PACKAGE_BUGREPORT "\n");
  exit(res);
}

void
showVersion()
{
  WRITE_MSG(1,
	    "lsxid " VERSION " -- shows the context which is associated to a file\n"
	    "This program is part of " PACKAGE_STRING "\n\n"
	    "Copyright (C) 2004 Enrico Scholz\n"
	    VERSION_COPYRIGHT_DISCLAIMER);
  exit(0);
}

void
fixupParams(struct Arguments UNUSED * args, int UNUSED argc)
{
  Vector_init(&xid_name_mapping, sizeof (struct XidNameMapping));
}

static int
cmpMap(void const *xid_v, void const *map_v)
{
  xid_t const * const			xid = xid_v;
  struct XidNameMapping const * const	map = map_v;

  return *xid - map->xid;
}

static char const *
lookupContext(xid_t xid)
{
  struct XidNameMapping	*res;

  res = Vector_search(&xid_name_mapping, &xid, cmpMap);

  if (res==0) {
    vcCfgStyle		style = vcCFG_AUTO;
    char const *	vcfg  = vc_getVserverByCtx(xid, &style, 0);

    res = Vector_insert(&xid_name_mapping, &xid, cmpMap);
    if (vcfg) res->name = vc_getVserverName(vcfg, style);
    else      res->name = 0;
    res->xid = xid;

    free(const_cast(char *)(vcfg));
  }

  assert(res!=0);
  
  return res->name;
}

static xid_t
getFileContext(char const *name)
{
  xid_t		res;
  uint32_t	mask = VC_IATTR_XID;
  
  if (vc_get_iattr(name, &res, 0, &mask)==-1)
    perror("vc_get_iattr()");

  return (mask&VC_IATTR_XID) ? res : VC_NOCTX;
}

bool
handleFile(char const *name, char const *display_name)
{
  xid_t		ctx = 0;
  char		buf[MAX(sizeof(ctx)*3+1, 20)];
  bool		need_write = true;

  memset(buf, ' ', sizeof buf);

#if 1
  ctx = getFileContext(name);
#else
#  warning Compiling in debug-code
  ctx = random() % 10 + 49213;
#endif
  
  if (ctx==VC_NOCTX) {
    memcpy(buf, "!!ERR!!", 7);
    Vwrite(1, buf, sizeof buf);
    need_write = false;
  }
  else if (global_args->do_mapping) {
    char const *	vname = lookupContext(ctx);
    if (!vname) buf[0] = '\0';
    else {
      size_t		l = strlen(vname);
      if (l<sizeof(buf)) Vwrite(1, buf, sizeof(buf)-l);
      Vwrite(1, vname, l);

      need_write = false;
    }
  }

  if (need_write) {
    size_t	l = utilvserver_fmt_ulong(buf, ctx);
    if (l<sizeof(buf)) Vwrite(1, buf+l, sizeof(buf)-l);
    Vwrite(1, buf, l);
  }

  Vwrite(1, "  ", 2);
  Vwrite(1, display_name, strlen(display_name));
  Vwrite(1, "\n", 1);

  return ctx!=VC_NOCTX;
}
